Refactor config - sync options in config file and command line

alanoe opened the pull request:
- Make options names consistent
- Sync options in config file and command line
- Rename config attributes to make them more consistent

Commit:
Move release notes specific options to a separate block in config.yaml

Commit:
Remove branch common option from config.yaml

This default branch was only used in the upgrade-version subcommand,
for which backwards compatibility is not needed. All current package
YAML configuration files have an explicit branch defined, so this
option is not needed anymore.

Commit:
Rename packages metadata repository attributes in config file and command line

New attributes names are more clearer and more consistent with other attrib names.

Commit:
Make subcommand names equal to config node names

Commit:
Rename default block in config.yaml to common

Commit:
Move discover_packages() method to outside config.py

config.py should handle only parsing command line and config YAML file,
nothing else.

Commit:
Create http_proxy parameter in config.yaml

Commit:
Improve build-package and release-notes subcommands names in host_os.py command line

- Rename build-package to build-packages. As subcommand supports
  building of N packages, new name makes more sense
- Rename release-notes to build-release-notes. New name is more
  consistent with other subcommands names such as build-packages
  and build-iso.

Commit:
Create build_packages block in config.yaml

Also, remove packages options where they are never used:
build-release-notes and update-versions-readme.

Also, create upgrade_versions node in config.yaml and duplicate the
packages option there to help development.

Commit:
Synchronize mock and push repo arguments in config.yaml and command line parser

Commit:
Make ISO creation get distro version specific Mock config file

This makes ISO creation consistent with packages creation.

Commit:
Rename package_group_list attrib in config.yaml to automated_install_packages_groups

New name is more intuitive.

Also, renamed related variables in ISO creator.

Commit:
Rename ISO creation attributes in config.yaml to make the names more intuitive

Commit:
Rename ISO builder Python module, class and variables names

"Spin" is Pungi terminology for building a custom ISO. We should not
propagate this anti-intuitive terminology in our code.

Commit:
Synchronize ISO build attributes in config file and host_os.py command line

Commit:
Stop using the global SUBCOMMANDS constant directly in ConfigParser class methods

Add subcommands param to _add_subparsers() and _setup_config_parser() instead of
using the variable defined outside the class directly. That restricts knowledge of
external variables to a single method in the class.

Also, document related methods.

Commit:
Reorganize config parser Python module

- Remove _add_subparsers(). This method is small, the calling method
  is small and this method is used only once, so there is no need to
  split it in a separate method.

- Improve methods documentation

- Improve methods names

Commit:
Rename upgrade-versions command line subcommand to update-versions

'update' is the terminology used in the rest of the UI and code.

Also, rename upgrade_versions attribute in config file to update_versions.

Commit:
Move selection of package builder and its conf file to BuildManager

Before, the selection of the package builder and its configuration file was
done by the Distribution class. This patch moves this reponsibility to the
facade BuildManager class and restricts the Distribution class responsibilities
to keeping the Linux distribution name, version, arch, endianness and checking
if it is supported by the build system. The purpose is to make the class design
simpler to understand.

Also, move PackageBuilder class to a separate Python module.

Commit:
Add log_size and verbose options to config.yaml

These options already exists in command line, but not in config file.

Commit:
Rename subparsers manager to make code easier to understand

Commit:
Rename add_subparser() method in config.py

This method adds several subparsers, not only one.

lisibann reviewed (changes requested):
These changes are not related to config refactoring (config file and command line). I suggest a separate PR.

alanoe commented:
LA> These changes are not related to config refactoring (config file and command line). I suggest a separate PR.

This will cause a big rebase effort, I vote against it. Why can't you just review each patch here? If the issue is that the PR description is incomplete, I can simply complete it.

Commit:
Update description of --result-dir argument in command line

As this argument is now used for RPMs build and ISO build, description
should be updated accordingly.

farosas commented:
start build

olavphibm commented on lib/iso_spinner.py:
```diff
@@ -29,21 +29,26 @@
+        self.work_dir = self.common_config.get('work_dir')
+        self.timestamp = datetime.datetime.now().isoformat()
+        self.result_dir = os.path.join(self.common_config.get('result_dir'),
+            'iso', self.timestamp)
```
Wrong indentation.

alanoe replied:
Done

olavphibm commented on lib/config.py:
```diff
@@ -83,16 +83,35 @@
+        dict(help='ISO name.',
+             default='OpenPOWER-Host_OS'),
+    ('--log-file',):
+        dict(help='ISO name.',
```
ISO log file path.

alanoe replied:
Done

olavphibm commented on lib/config.py:
```diff
@@ -83,16 +83,35 @@
+    ('--automated-install-packages-groups',):
+        dict(help='Packages and packages groups which are installed using automated installation', nargs='*'),
+    ('--mock-iso-repo-name',):
+        dict(help='Name of the yum repository which will be created with OpenPOWER Host OS packages'),
```
I'd replace "which will be created" with "to create"

alanoe replied:
Done

olavphibm commented on lib/config.py:
```diff
@@ -83,16 +83,35 @@
+    ('--mock-iso-repo-name',):
+        dict(help='Name of the yum repository which will be created with OpenPOWER Host OS packages'),
+    ('--mock-iso-repo-dir',):
+        dict(help='Directory path of the yum repository which will be created with OpenPOWER Host OS packages'),
```
I'd replace "which will be created" with "to create"

alanoe replied:
Done

olavphibm commented on lib/build_manager.py:
```diff
@@ -15,11 +15,13 @@
 
 import logging
 
+import lib.centos
```
This should be after the last `from`

alanoe replied:
Done

olavphibm commented on lib/iso_spinner.py:
```diff
@@ -29,21 +29,26 @@
         try:
             utils.run_command("%s -r %s %s %s" % (
-                self.mock_binary, self.config.get('mock_config'),
+                self.mock_binary, distro.config_file,
```
This attribute does not exist in `LinuxDistribution` class. It is just a local variable in `CentOS` constructor.

alanoe replied:
Done

olavphibm commented on lib/iso_builder.py:
```diff
@@ -46,9 +46,11 @@ def _run_mock_command(self, cmd):
             self.common_config.get('distro_version'),
             self.common_config.get('arch_and_endianness'))
+        mock_config_file = self.config.get('mock_config').get(distro.lsb_name).get(
+            distro.version)
```
This should probably be in the previous patch "Make ISO creation get distro version specific Mock config file "

olavphibm commented on tests/unit/config_tests.py:
```diff
@@ -51,7 +51,7 @@ def test_parse_arguments_list_WithLongArgument_ShouldParseArgumentValue(self, ar
         (['build-release-notes'], 'push_repo_branch', 'master'),
         (['build-release-notes'], 'updater_name', None),
         (['build-release-notes'], 'updater_email', None),
-        (['build-iso'], 'packages_dir', './result'),
```
This was mistakenly changed in a previous patch of this PR.

alanoe commented on lib/packages_manager.py:
```diff
@@ -28,6 +30,7 @@ def __init__(self, packages_names):
         self.packages_names = packages_names
         self.packages = []
 
+
```
Done

alanoe commented on lib/config.py:
```diff
@@ -263,9 +264,11 @@ def parse(self):
-            if key in config['iso']:
-                config['iso'][key] = value
-                args.pop(key)
+            for node in COMMAND_TO_CONFIG_NODE.values():
```
Both will produce the same result. You are right, the for is unnecessary. Done

alanoe commented on lib/config.py:
```diff
@@ -241,7 +241,7 @@ def parse(self):
-            "build-iso": "iso",
+            "build-iso": "build_iso",
             "release-notes": "release_notes"
         }
```
This dict will be deleted in a future patch, I prefer to not touch it here.

alanoe commented on tools/upgrade_versions.py:
```diff
@@ -225,7 +225,7 @@ def push_packages_head_commit(
 def run(CONF):
     versions_repo = setup_versions_repository(CONF)
-    packages_to_update = CONF.get('common').get('packages') or PACKAGES
+    packages_to_update = PACKAGES
```
Not sure I follow. 

1) What is the purpose of testing with only a subset? It may work with the subset and break with a specific package outside the subset, we should test with all packages.
2) Do you have a subset of packages (Git/SVN) repositories in your dev environment? I have all in mine so I can test anything I want in the build system.

olavphibm replied:
1. It is faster. Not every test during development needs to be complete.
2. Sometimes I need to clean the environment to test a new feature, sometimes the paths change (gyr's latest patches), sometimes I just don't want to wait to download the last week's of kernels updates.

alanoe commented on config.yaml:
```diff
@@ -19,6 +19,10 @@ common:
+build_packages:
+ packages:
+ keep_build_dir: False
+ result_dir:  "./result"
```
Already fixed

mopsfelder commented on config.yaml:
```diff
@@ -22,6 +20,9 @@ default:
  force_rebuild: False
+release_notes:
+ release_notes_repo_url: "https://github.com/open-power-host-os/open-power-host-os.github.io.git"
+ release_notes_repo_branch: "master"
```
Not related to this change but do we need to repeat the name of the root node in its elements?  For example, `relete_notes_repo_url` is inside `release_notes`, it could be just `repo_url`.  Something like this:

```yaml
release_notes:
  repo_url:
  repo_branch:
```

olavphibm replied:
@alanoe is making all options here match the same name as in `config.py`. Having a `repo_url` command line option there would be ambiguous. I believe this can be improved in future PRs, when options will be properly separated among commands.

mopsfelder reviewed (commented):
+1 for f345dab6cfcf56d11734244cfec96242b8b88de4

lisibann commented:
I reviewed all commits related to this PR. The comment was related to "Move selection of package builder and its conf file to BuildManager". No problem to keep it as is.
